src/v/cluster/metadata_cache.cc

```diff
@@ -248,4 +276,47 @@ metadata_cache::get_default_shadow_indexing_mode() const {
     }
     return m;
 }
+
+ss::future<> metadata_cache::broadcast_disk_space_state(
+  const storage::disk_space_alert state) {
+    return container().invoke_on_others(
+      [state](metadata_cache& mc) { mc.set_disk_space_state(state); });
+}
+
+void metadata_cache::set_disk_space_state(const storage::disk_space_alert s) {
+    _disk_space_state = s;
+    vlog(clusterlog.debug, "set_disk_space_state {}", s);
+}
+
+bool metadata_cache::should_reject_writes() const {
+    return _disk_space_state == storage::disk_space_alert::degraded;
```

Comment:
What about the situations in which the node that is accepting writes (partition leader) has plenty of space available but partition replicas are almost out of disk space, will we block writes in this situation ?

Reply:
This is the specific design of v1. Check out the RFC if you haven't already. It need to be improved in the future but this is a big improvement, and it is configurable if we run into imbalance isues.